### PR TITLE
Feature/s4hana new variables

### PR DESCRIPTION
**Pull request by arbulu89**
Implement the usage of S/4HANA 1909.
Some items:
- Tune the pillar files almost completely from terraform variables. I think this is the best choice as passing some variables and not other is causing confusion. By now, this is the best approach in my opinion. We can find a better code solution maybe later to avoid having so many code duplicated.
- Adapt the pillar files to use terraform variables data
- Adapt the pillar files to install S/4HANA. I don't have any documentation about the used new variables in the pillar file. It has been a long try/fail process, to find out which are the correct parameters, as there is not any information in the internet unfortunately.
- Adapt the pillar files to deploy properly the non HA scenarios, and scenarios where the NW ASCS and PAS are running in the same machine (apparently a quite standard scenario in small organizations).

New variables for the user:
- hana_sid
- hana_cost_optimized_sid
- hana_instance_number
- hana_cost_optimized_instance_number
- hana_master_password
- hana_cost_optimized_master_password
- hana_primary_site
- hana_secondary_site
- netweaver_app_server_count
- netweaver_ascs_instance_number
- netweaver_ers_instance_number
- netweaver_pas_instance_number
- netweaver_master_password

Depends on:
https://github.com/SUSE/salt-shaptools/pull/73
https://github.com/SUSE/sapnwbootstrap-formula/pull/67/

### CODE REVIEW

File: libvirt/variables.tf [hunk 1]
```diff
@@ -145,6 +145,12 @@ variable "hana_node_disk_size" {
   default     = 68719476736
 }
 
+variable "hana_fstype" {
```
Comment (Simranpal) [inline libvirt/variables.tf:148]: was this variable missing in `libvirt`?

File: libvirt/modules/netweaver_node/salt_provisioner.tf [hunk 2]
```diff
@@ -27,8 +27,16 @@ host_ips: [${join(", ", formatlist("'%s'", var.host_ips))}]
 virtual_host_ips: [${join(", ", formatlist("'%s'", var.virtual_host_ips))}]
 host_ip: ${element(var.host_ips, count.index)}
 hana_ip: ${var.hana_ip}
+hana_sid: ${var.hana_sid}
+hana_instance_number: ${var.hana_instance_number}
+hana_master_password: ${var.hana_master_password}
 ha_enabled: ${var.ha_enabled}
+app_server_count: ${var.app_server_count}
```
[the remark this answers is not in the document]
Comment (Simranpal) [reply libvirt/modules/netweaver_node/salt_provisioner.tf:34]: i checked and it is not used on grains so far

**Review by Simranpal — APPROVED** [review]
Thanks @arbulu89 I have some minor suggestions in mainly documentation.
Besides that I did not understand the usage of xscs server.

File: aws/terraform.tfvars.example [hunk 3]
```diff
@@ -175,6 +175,20 @@ hana_inst_master = "s3://sapdata/sap_inst_media/51053381"
...
+# HANA instance configuration
+# Find some references about the variables in:
+# https://help.sap.com/doc/e9702d76c3284623b02de196c0e79e49/2.0.05/en-US/SAP_HANA_Server_Installation_Guide_en.pdf
+# HANA instance system identifier. It's composed by 3 characters string
+#hana_sid = "prd"
+# HANA instance number. It's composed by 2 integers string
+#hana_instance_number = "00"
+# HANA instance master password
+#hana_master_password = "YourPassword1234"
+# HANA primary site name. Only used if system replication or HA capabilities are enabled
```
[the remark this answers is not in the document]
Comment (arbulu89) [reply aws/terraform.tfvars.example:187]: Ok. As system replication and HA are always enabled together, I wanted to emphasize this, but you are right

File: aws/terraform.tfvars.example [hunk 4]
```diff
@@ -245,6 +259,13 @@ hana_inst_master = "s3://sapdata/sap_inst_media/51053381"
 #############################
 
 #netweaver_enabled = true
+
+# Netweaver APP server count (PAS and AAS)
+# Set to 0 to install the PAS instance in the same instance than the ASCS. This means only 1 machine is installed in the deployment (2 if HA capabilities are enabled)
```
[the remark this answers is not in the document]
Comment (arbulu89) [reply aws/terraform.tfvars.example:264]: Yes. It looks like the simplest deployments consist of a single HANA database, and a single NW or S/4HANA machine with ASCS and PAS running there, both in the same.
I will correct the typo

File: libvirt/main.tf [hunk 5]
```diff
@@ -29,10 +29,13 @@ locals {
   drbd_ips         = length(var.drbd_ips) != 0 ? var.drbd_ips : [for ip_index in range(local.drbd_ip_start, local.drbd_ip_start + 2) : cidrhost(local.iprange, ip_index)]
   drbd_cluster_vip = var.drbd_cluster_vip != "" ? var.drbd_cluster_vip : cidrhost(local.iprange, local.drbd_ip_start + 2)
 
+  netweaver_xscs_server_count = var.netweaver_enabled ? (var.netweaver_ha_enabled ? 2 : 1) : 0
```
[the remark this answers is not in the document]
Comment (arbulu89) [reply libvirt/main.tf:32]: The `xscs` servers are the `ascs` and `ers`. Depending on the provider (and in the future most probably we will need to port this to all of them), the ASCS and ERS machines must be more powerful than the PAS/AAS machines. This variable is used to know how many of these machines are of this type. By now, this logic is only implemented in azure, but for that, we need to provide the `netweaver` module this information.

Actually, maybe it's redundant, as the `netweaver_ha_enabled` is already passed, and it's based only in this parameter. But it helps to make the code a bit cleaner in the other hand

File: libvirt/modules/netweaver_node/salt_provisioner.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply libvirt/modules/netweaver_node/salt_provisioner.tf:34]: No, it's not needed. It's only used to identify the number of `xscs` (ascs/ers) the deployment must create using their own configuration (instance type, network configuration, etc).
Basically, the infrastructure for them is different

File: libvirt/terraform.tfvars.example [hunk 6]
```diff
@@ -56,6 +56,20 @@ hana_inst_media = "url-to-your-nfs-share:/sapdata/sap_inst_media/51053381"
 # Set the IP addresses for the HANA machines. Leave this commented to get an autogenerated addresses
 #hana_ips         = ["192.168.XXX.Y", "192.168.XXX.Y+1"]
 
+# HANA instance configuration
+# Find some references about the variables in:
+# https://help.sap.com/doc/e9702d76c3284623b02de196c0e79e49/2.0.05/en-US/SAP_HANA_Server_Installation_Guide_en.pdf
```
[the remark this answers is not in the document]
Comment (arbulu89) [reply libvirt/terraform.tfvars.example:61]: Ok, it's safer to point to the generic link

File: libvirt/variables.tf (hunk 1 above)
Comment (arbulu89) [reply libvirt/variables.tf:148]: No, it just moved from one line in the code to another, to have more logic

File: libvirt/main.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (Simranpal) [reply libvirt/main.tf:32]: okay, thanks, it makes sense

**Pull request merged**